tfhe-zk-pok/src/proofs/mod.rs

```diff
@@ -174,3 +174,202 @@ pub mod pke;
 pub mod pke_v2;
 pub mod range;
 pub mod rlwe;
+
+#[cfg(test)]
+mod test {
+    #![allow(non_snake_case)]
+    use std::fmt::Display;
+
+    use bincode::ErrorKind;
+    use rand::rngs::StdRng;
+    use rand::Rng;
+    use serde::{Deserialize, Serialize};
+
+    use crate::curve_api::Compressible;
+
+    // One of our usecases uses 320 bits of additional metadata
+    pub(super) const METADATA_LEN: usize = (320 / u8::BITS) as usize;
+
+    pub(super) enum Compress {
+        Yes,
+        No,
+    }
+
+    pub(super) fn serialize_then_deserialize<
+        Params: Compressible + Serialize + for<'de> Deserialize<'de>,
+    >(
+        public_params: &Params,
+        compress: Compress,
+    ) -> bincode::Result<Params>
+    where
+        <Params as Compressible>::Compressed: Serialize + for<'de> Deserialize<'de>,
+        <Params as Compressible>::UncompressError: Display,
+    {
+        match compress {
+            Compress::Yes => Params::uncompress(bincode::deserialize(&bincode::serialize(
+                &public_params.compress(),
+            )?)?)
+            .map_err(|e| Box::new(ErrorKind::Custom(format!("Failed to uncompress: {}", e)))),
+            Compress::No => bincode::deserialize(&bincode::serialize(&public_params)?),
+        }
+    }
+
+    pub(super) fn polymul_rev(a: &[i64], b: &[i64]) -> Vec<i64> {
+        assert_eq!(a.len(), b.len());
+        let d = a.len();
+        let mut c = vec![0i64; d];
+
+        for i in 0..d {
+            for j in 0..d {
+                if i + j < d {
+                    c[i + j] = c[i + j].wrapping_add(a[i].wrapping_mul(b[d - j - 1]));
+                } else {
+                    c[i + j - d] = c[i + j - d].wrapping_sub(a[i].wrapping_mul(b[d - j - 1]));
+                }
+            }
+        }
+
+        c
+    }
+
+    #[derive(Copy, Clone)]
+    pub(super) struct PkeTestParameters {
+        pub(super) d: usize,
+        pub(super) k: usize,
+        pub(super) B: u64,
+        pub(super) q: u64,
+        pub(super) t: u64,
+        pub(super) msbs_zero_padding_bit_count: u64,
+    }
+
+    pub(super) const DEFAULT_TEST_PKE_PARAMS: PkeTestParameters = PkeTestParameters {
+        d: 2048,
```

Comment:
not the correct d, you want the one from tfhe/src/shortint/parameters/compact_public_key_only/p_fail_2_minus_64/ks_pbs.rs:9